theano/tensor/nnet/corr.py

```diff
@@ -66,20 +67,29 @@ def __init__(self, border_mode="valid", subsample=(1, 1),
                 raise ValueError(
                     'invalid border_mode {}, which must be a '
                     'non-negative integer'.format(border_mode))
-            border_mode = (border_mode, border_mode)
-        if isinstance(border_mode, tuple):
-            if len(border_mode) != 2 or border_mode[0] < 0 or border_mode[1] < 0:
+            border_mode = ((border_mode, border_mode),) * 2
+        elif isinstance(border_mode, tuple):
+            if len(border_mode) != 2:
                 raise ValueError(
-                    'invalid border_mode {}, which must be a '
-                    'pair of non-negative integers'.format(border_mode))
-            pad_h, pad_w = map(int, border_mode)
-            border_mode = (pad_h, pad_w)
-        if not ((isinstance(border_mode, tuple) and min(border_mode) >= 0) or
-                border_mode in ('valid', 'full', 'half')):
+                    'invalid border_mode {} which must be a '
+                    'tuple of length 2'.format(border_mode))
+            border = ()
+            for mode in border_mode:
+                if isinstance(mode, integer_types) and mode >= 0:
+                    border += ((mode, mode),)
+                elif isinstance(mode, tuple) and len(mode) == 2 and \
+                        min(mode) >= 0:
+                    border += ((int(mode[0]), int(mode[1])),)
```

Comment:
I'm not really sure why those casts to `int` were there in the original version (`map(int, border_mode)`). But if you want to keep it backwards compatible, your new version is changing the way that non-integers are handled. It would be better to swap the `if`
```python
if isinstance(mode, tuple):
    # ...
else:
    # mode must be a number, but has to be casted to an int
```

Reply:
Can't it also be something else in that case, say a string or a list?

Reply:
But I see that the original code doesn't account for that either. Okay then